app/helpers/agent_helper.rb

```diff
@@ -57,4 +57,39 @@ def agent_dry_run_with_event_mode(agent)
       'maybe'.freeze
     end
   end
+
+  def agent_type_icon(agent, agents)
+    receiver_count = links_counter_cache(agents)[:links_as_receiver][agent.id] || 0
+    control_count  = links_counter_cache(agents)[:control_links_as_controller][agent.id] || 0
+    source_count   = links_counter_cache(agents)[:links_as_source][agent.id] || 0
+
+    if control_count > 0 && receiver_count > 0
+      content_tag ('span') do
+        concat icon_tag('glyphicon-arrow-right')
+        concat tag('br')
+        concat icon_tag('glyphicon-log-out', class: 'glyphicon-flipped')
+      end
+    elsif control_count > 0 && receiver_count == 0
+      icon_tag('glyphicon-log-out', class: 'glyphicon-flipped')
+    elsif receiver_count > 0 && source_count == 0
+      icon_tag('glyphicon-arrow-right')
+    elsif receiver_count == 0 && source_count > 0
+      icon_tag('glyphicon-arrow-left')
+    elsif receiver_count > 0 && source_count > 0
+      icon_tag('glyphicon-transfer')
+    else
+      icon_tag('glyphicon-unchecked')
+    end
+  end
+
+  private
+
+  def links_counter_cache(agents)
+    @counter_cache ||= {}.tap do |cache|
```

Comment:
The cache would not be polluted because it's only set once per request, but calling it with a different `agents` array it could incorrectly return `nil` if the agents was not present in the first array. I added `agents.__id__` as an additional cache key.